File: scripts/migrations/migrate.ts
#!/usr/bin/env node
/**
 * Migration Runner
 *
 * Usage:
 *   pnpm migrate:up          # Apply all pending migrations
 *   pnpm migrate:down        # Rollback last migration
 *   pnpm migrate:status      # Show current version
 *   pnpm migrate:history     # Show migration history
 *
 * Addresses: Risk #1 from ARCHITECTURE_REVIEW_2026-01-21.md
 */

import { readdir, readFile } from 'fs/promises';
import { join, basename } from 'path';
import { createHash } from 'crypto';
import { logger } from '@quantbot/utils';

// For now, use simple DuckDB operations until we export schema-version from @quantbot/storage
import duckdb from 'duckdb';

interface MigrationFile {
  version: number;
  name: string;
  filepath: string;
  checksum: string;
  upSql: string;
  downSql: string | null;
}

async function loadMigrations(migrationsDir: string): Promise<MigrationFile[]> {
  const files = await readdir(migrationsDir);
  const sqlFiles = files.filter((f) => f.endsWith('.sql') && f.match(/^\d{3}_/));

  const migrations: MigrationFile[] = [];

  for (const file of sqlFiles) {
    const filepath = join(migrationsDir, file);
    const content = await readFile(filepath, 'utf-8');

    // Extract version from filename (e.g., "001_add_column.sql" → 1)
    const version = parseInt(file.substring(0, 3), 10);
    const name = file.substring(4).replace('.sql', '');

    // Calculate checksum
    const checksum = createHash('sha256').update(content).digest('hex');

    // Parse up/down migrations (simple convention: split on "-- Rollback:")
    const parts = content.split(/-- Rollback:|-- Down:/i);
    const upSql = parts[0]?.trim() ?? '';
    const downSql = parts[1]?.trim() || null;

    migrations.push({
      version,
      name,
      filepath,
      checksum,
      upSql,
      downSql,
    });
  }

  return migrations.sort((a, b) => a.version - b.version);
}

async function migrateUp(duckdbPath: string): Promise<void> {
  console.log('🔄 Running migrations...\n');

  const db = new DuckDBClient({ dbPath: duckdbPath });

  try {
    // Initialize tracking table
    await initializeSchemaTracking(db as any);

    // Load all migration files
    const migrationsDir = join(process.cwd(), 'packages/storage/migrations');
    const migrations = await loadMigrations(migrationsDir);

    // Get current version
    const currentVersion = await getCurrentSchemaVersion(db as any, 'duckdb');
    console.log(`📌 Current version: ${currentVersion}`);

    // Find pending migrations
    const pending = migrations.filter((m) => m.version > currentVersion);

    if (pending.length === 0) {
      console.log('✅ No pending migrations');
      return;
    }

    console.log(`📦 Found ${pending.length} pending migration(s)\n`);

    // Apply each migration
    for (const migration of pending) {
      console.log(`⏳ Applying migration ${migration.version}: ${migration.name}...`);

      const startTime = Date.now();

      try {
        // Execute migration SQL
        await (db as any).run(migration.upSql);

        const executionTime = Date.now() - startTime;

        // Record migration
        await recordMigration(db as any, {
          version: migration.version,
          name: migration.name,
          description: `Migration from file: ${basename(migration.filepath)}`,
          database_type: 'duckdb',
          checksum: migration.checksum,
          execution_time_ms: executionTime,
          rollback_sql: migration.downSql,
        });

        console.log(`✅ Migration ${migration.version} applied (${executionTime}ms)\n`);
      } catch (error) {
        console.error(`❌ Migration ${migration.version} failed:`, error);
        throw error;
      }
    }

    const newVersion = await getCurrentSchemaVersion(db as any, 'duckdb');
    console.log(`\n🎉 Migrations complete! Version: ${currentVersion} → ${newVersion}`);
  } finally {
    // Note: DuckDBClient might need explicit close method
    // await db.close();
  }
}

async function migrateDown(duckdbPath: string): Promise<void> {
  console.log('🔄 Rolling back last migration...\n');

  const db = new DuckDBClient({ dbPath: duckdbPath });

  try {
    const currentVersion = await getCurrentSchemaVersion(db as any, 'duckdb');

    if (currentVersion === 0) {
      console.log('⚠️  Already at version 0, cannot rollback further');
      return;
    }

    // Get the migration to rollback
    const history = await getMigrationHistory(db as any, 'duckdb');
    const lastMigration = history.find(
      (m) => m.version === currentVersion && m.status === 'applied'
    );

    if (!lastMigration) {
      console.error(`❌ Migration ${currentVersion} not found or not applied`);
      return;
    }

    if (!lastMigration.rollback_sql) {
      console.error(`❌ Migration ${currentVersion} has no rollback SQL`);
      return;
    }

    console.log(`⏳ Rolling back migration ${currentVersion}: ${lastMigration.name}...`);

    const startTime = Date.now();

    try {
      // Execute rollback SQL
      await (db as any).run(lastMigration.rollback_sql);

      const executionTime = Date.now() - startTime;

      // Mark as rolled back
      await recordRollback(db as any, currentVersion, 'duckdb');

      console.log(`✅ Migration ${currentVersion} rolled back (${executionTime}ms)\n`);

      const newVersion = await getCurrentSchemaVersion(db as any, 'duckdb');
      console.log(`🎉 Rollback complete! Version: ${currentVersion} → ${newVersion}`);
    } catch (error) {
      console.error(`❌ Rollback ${currentVersion} failed:`, error);
      throw error;
    }
  } finally {
    // await db.close();
  }
}

async function showStatus(duckdbPath: string): Promise<void> {
  const db = new DuckDBClient({ dbPath: duckdbPath });

  try {
    await initializeSchemaTracking(db as any);

    const currentVersion = await getCurrentSchemaVersion(db as any, 'duckdb');
    const history = await getMigrationHistory(db as any, 'duckdb');

    console.log('📊 Schema Status\n');
    console.log(`Current version: ${currentVersion}`);
    console.log(`Total migrations: ${history.length}`);
    console.log(`Applied: ${history.filter((m) => m.status === 'applied').length}`);
    console.log(`Rolled back: ${history.filter((m) => m.status === 'rolled_back').length}\n`);
  } finally {
    // await db.close();
  }
}

async function showHistory(duckdbPath: string): Promise<void> {
  const db = new DuckDBClient({ dbPath: duckdbPath });

  try {
    await initializeSchemaTracking(db as any);

    const history = await getMigrationHistory(db as any, 'duckdb');

    console.log('📜 Migration History\n');
    console.table(
      history.map((m) => ({
        Version: m.version,
        Name: m.name,
        Status: m.status,
        Applied: m.applied_at.toISOString().substring(0, 19),
        Time: m.execution_time_ms ? `${m.execution_time_ms}ms` : '-',
      }))
    );
  } finally {
    // await db.close();
  }
}

// CLI
const command = process.argv[2];
const duckdbPath = process.env.DUCKDB_PATH || 'data/quantbot.duckdb';

(async () => {
  try {
    switch (command) {
      case 'up':
        await migrateUp(duckdbPath);
        break;
      case 'down':
        await migrateDown(duckdbPath);
        break;
      case 'status':
        await showStatus(duckdbPath);
        break;
      case 'history':
        await showHistory(duckdbPath);
        break;
      default:
        console.log(`
Usage: pnpm migrate <command>

Commands:
  up        Apply all pending migrations
  down      Rollback last migration
  status    Show current schema version
  history   Show migration history

Environment:
  DUCKDB_PATH   Path to DuckDB file (default: data/quantbot.duckdb)
        `);
        process.exit(1);
    }
  } catch (error) {
    console.error('❌ Migration failed:', error);
    process.exit(1);
  }
})();

File: tools/storage/migrate.py
#!/usr/bin/env python3
"""
Schema Migration Runner

Usage:
    python tools/storage/migrate.py up [--duckdb PATH]
    python tools/storage/migrate.py status [--duckdb PATH]
    python tools/storage/migrate.py history [--duckdb PATH]

Addresses: Risk #1 from ARCHITECTURE_REVIEW_2026-01-21.md
          "Schema migration strategy is implicit"
"""

import sys
import os
import argparse
from pathlib import Path
from datetime import datetime
import hashlib
import duckdb

def load_migrations(migrations_dir: Path) -> list:
    """Load all migration SQL files"""
    migrations = []
    
    if not migrations_dir.exists():
        return migrations
    
    for file in sorted(migrations_dir.glob("*.sql")):
        if not file.name[0].isdigit():
            continue
            
        version = int(file.name[:3])
        name = file.name[4:-4]  # Remove "000_" and ".sql"
        sql = file.read_text()
        checksum = hashlib.sha256(sql.encode()).hexdigest()
        
        migrations.append({
            'version': version,
            'name': name,
            'filepath': str(file),
            'sql': sql,
            'checksum': checksum
        })
    
    return migrations

def get_current_version(conn: duckdb.DuckDBPyConnection, db_type: str = 'duckdb') -> int:
    """Get current schema version"""
    try:
        result = conn.execute("""
            SELECT MAX(version) as current_version
            FROM schema_migrations
            WHERE database_type = ?
              AND status = 'applied'
        """, [db_type]).fetchone()
        
        return result[0] if result and result[0] is not None else 0
    except Exception:
        # Table doesn't exist yet
        return 0

def migrate_up(duckdb_path: str):
    """Apply all pending migrations"""
    print("🔄 Running migrations...\n")
    
    conn = duckdb.connect(duckdb_path)
    
    try:
        current_version = get_current_version(conn)
        print(f"📌 Current version: {current_version}")
        
        # Load migration files
        migrations_dir = Path("packages/storage/migrations")
        migrations = load_migrations(migrations_dir)
        
        # Filter pending migrations (include current_version if it's 0 and migration 0 exists)
        if current_version == 0:
            # First run: apply migration 0 and all others
            pending = migrations
        else:
            # Normal: only apply migrations > current_version
            pending = [m for m in migrations if m['version'] > current_version]
        
        if not pending:
            print("✅ No pending migrations")
            return
        
        print(f"📦 Found {len(pending)} pending migration(s)\n")
        
        for migration in pending:
            print(f"⏳ Applying migration {migration['version']}: {migration['name']}...")
            
            start_time = datetime.now()
            
            try:
                # Execute migration
                conn.execute(migration['sql'])
                
                execution_time = int((datetime.now() - start_time).total_seconds() * 1000)
                
                # Record migration in tracking table
                # Note: Migration 000 creates the schema_migrations table itself
                # So we only try to record if version > 0
                try:
                    conn.execute("""
                        INSERT INTO schema_migrations 
                        (version, name, description, database_type, checksum, execution_time_ms, status)
                        VALUES (?, ?, ?, 'duckdb', ?, ?, 'applied')
                    """, [
                        migration['version'],
                        migration['name'],
                        f"Migration: {migration['name']}",
                        migration['checksum'],
                        execution_time
                    ])
                except Exception as record_error:
                    # If version 0, the table might not exist yet (it's being created by this migration)
                    if migration['version'] == 0:
                        print(f"   (Schema migrations table created by this migration)")
                    else:
                        raise record_error
                
                print(f"✅ Migration {migration['version']} applied ({execution_time}ms)\n")
                
            except Exception as e:
                print(f"❌ Migration {migration['version']} failed: {e}")
                raise
        
        new_version = get_current_version(conn)
        print(f"\n🎉 Migrations complete! Version: {current_version} → {new_version}")
        
    finally:
        conn.close()

def show_status(duckdb_path: str):
    """Show current schema status"""
    conn = duckdb.connect(duckdb_path, read_only=True)
    
    try:
        current_version = get_current_version(conn)
        
        print("\n📊 Schema Status")
        print(f"Database: {duckdb_path}")
        print(f"Current version: {current_version}")
        
        # Try to get migration count
        try:
            result = conn.execute("""
                SELECT 
                    COUNT(*) as total,
                    SUM(CASE WHEN status = 'applied' THEN 1 ELSE 0 END) as applied
                FROM schema_migrations
                WHERE database_type = 'duckdb'
            """).fetchone()
            
            if result:
                print(f"Total migrations: {result[1]} applied")
        except:
            pass
        
        print()
        
    finally:
        conn.close()

def show_history(duckdb_path: str):
    """Show migration history"""
    conn = duckdb.connect(duckdb_path, read_only=True)
    
    try:
        rows = conn.execute("""
            SELECT version, name, status, applied_at, execution_time_ms
            FROM schema_migrations
            WHERE database_type = 'duckdb'
            ORDER BY version
        """).fetchall()
        
        print("\n📜 Migration History\n")
        print(f"{'Ver':<5} {'Name':<30} {'Status':<12} {'Applied':<20} {'Time':<10}")
        print("=" * 80)
        
        for row in rows:
            version, name, status, applied_at, exec_time = row
            exec_time_str = f"{exec_time}ms" if exec_time else "-"
            applied_str = str(applied_at)[:19] if applied_at else "-"
            print(f"{version:<5} {name:<30} {status:<12} {applied_str:<20} {exec_time_str:<10}")
        
        print()
        
    except Exception as e:
        print(f"⚠️  No migration history available: {e}\n")
    finally:
        conn.close()

def main():
    parser = argparse.ArgumentParser(description='Schema Migration Runner')
    parser.add_argument('command', choices=['up', 'status', 'history'],
                       help='Migration command')
    parser.add_argument('--duckdb', default=os.environ.get('DUCKDB_PATH', 'data/quantbot.duckdb'),
                       help='Path to DuckDB file')
    
    args = parser.parse_args()
    
    if args.command == 'up':
        migrate_up(args.duckdb)
    elif args.command == 'status':
        show_status(args.duckdb)
    elif args.command == 'history':
        show_history(args.duckdb)

if __name__ == '__main__':
    main()

